src/briefcase/integrations/subprocess.py

```diff
@@ -145,6 +146,107 @@ def inner(sub: Subprocess, args: SubprocessArgsT, *wrapped_args, **wrapped_kwarg
     return inner
 
 
+class PopenOutputStreamer(threading.Thread):
+    def __init__(
+        self,
+        label: str,
+        popen_process: subprocess.Popen,
+        logger: Log,
+        capture_output: bool = False,
+        filter_func: Callable[[str], Iterator[str]] | None = None,
+    ):
+        """Thread for streaming stdout for a Popen process.
+
+        :param label: Descriptive name for process being streamed
+        :param popen_process: Popen process with stdout to stream
+        :param logger: logger for printing to console
+        :param capture_output: Retain process output in ``output_queue`` via a
+            ``queue.Queue`` instead of printing to console
+        :param filter_func: a callable that will be invoked on every line of output
+            that is streamed; see ``Subprocess.stream_output`` for details
+        """
+        super().__init__(name=f"{label} output streamer", daemon=True)
+
+        self.popen_process = popen_process
+        self.logger = logger
+        self.capture_output = capture_output
+        self.filter_func = filter_func
+
+        # arbitrarily large maxsize to prevent unbounded memory use if things go south
+        self.output_queue = queue.Queue(maxsize=10_000_000)
+        self.stop_flag = threading.Event()
+
+    def run(self):
+        """Stream output for a Popen process."""
+        try:
+            while output_line := self._readline():
+                if not self.stop_flag.is_set():
+                    filtered_output, stop_streaming = self._filter(output_line)
+
+                    for filtered_line in filtered_output:
+                        if self.capture_output:
+                            self.output_queue.put_nowait(filtered_line)
+                        else:
+                            self.logger.info(filtered_line)
+
+                    if stop_streaming:
+                        self.stop_flag.set()
+
+                if self.stop_flag.is_set():
```

Comment:
Is there a reason for evaluating `not is_set` then `is set` as separate if statements? Naïvely, it feels like this should be an if/else; but I can imagine there might be a complex interaction with `readline()` that I'm missing. If there is, there's a need for a docstring; if there isn't - it's an easy simplification :-)

Reply:
This is intentional because I didn't want `filter_func()` to set the flag and then come back around for another `readline()` where the streamer could remain stuck if there's nothing in the `stdout` buffer. So, this way, whether the stop flag was set either while `readline()` was waiting or as a consequence of `filter_func()`, the thread would then exit without any further reading or printing/capturing.

I can add something to encompass this....unless you might see it differently.

Reply:
That's exactly the sort of edge case I was thinking of. It makes perfect sense; we just need a docstring so that future travellers don't look at the code and thing "oh, I can clean this up!".